### PR TITLE
rgw: prevent spurious/lost notifications in the index completion thread

### PR DESCRIPTION
**Pull request by cbodley**
cbodley: cherry-picked from the wip-rgw-multisite-reshard feature branch after merge in https://github.com/ceph/ceph/pull/45131, so it can be backported to earlier releases. removed logging changes in `guard_reshard()` due to conflicts with commit https://github.com/ceph/ceph/commit/85848c02f4f52803cb2c7108ebad58422a9c6cc4 that's specific to multisite reshard

Fixes: https://tracker.ceph.com/issues/54435

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by yuvalif — APPROVED** [review]

**Comment by ljflores** [comment]
jenkins test make check

**Comment by cbodley** [comment]
@yuvalif seeing some UninitCondition and SyscallParam valgrind errors from this one in https://pulpito.ceph.com/cbodley-2022-03-04_14:14:50-rgw-wip-cbodley-testing-distro-default-smithi/

they're all under `RGWIndexCompletionManager::process()`, and mostly relate to its mutex and condition variable

File: src/rgw/rgw_rados.cc [hunk 1]
```diff
@@ -804,121 +804,65 @@ struct complete_op_data {
...
+  const int num_shards;
+  ceph::containers::tiny_vector<ceph::mutex> locks;
+  std::vector<set<complete_op_data*>> completions;
+  std::vector<complete_op_data*> retry_completions;
 
-  for (auto c : comps) {
-    std::unique_ptr<complete_op_data> up{c};
+  std::thread retry_thread;
+  std::condition_variable cond;
+  std::mutex retry_completions_lock;
```
Comment (cbodley) [inline src/rgw/rgw_rados.cc:816]: it may just be that `retry_thread`'s constructor is being called before its `cond` and `retry_completions_lock` are?

**Comment by cbodley** [comment]
rerun was clean https://pulpito.ceph.com/cbodley-2022-03-07_22:02:15-rgw-wip-cbodley-testing-distro-default-smithi/

**Pull request merged**